### ISSUE TITLE
sudo make install fails

### ISSUE THREAD

**Issue by magick93** [grpc/grpc#12405]
### Should this be an issue in the gRPC issue tracker?
 Yes
 
### What version of gRPC and what language are you using?
 Master branch, 1.0
 
### What operating system (Linux, Windows, …) and version?
 Ubuntu 17.0
 
### What runtime / compiler are you using (e.g. python version or version of gcc)

gcc --version
gcc (Ubuntu 6.3.0-12ubuntu2) 6.3.0 20170406

 
### What did you do?

```
 $ git clone -b $(curl -L https://grpc.io/release) https://github.com/grpc/grpc
 $ cd grpc
 $ git submodule update --init
 $ make
 $ [sudo] make install < - fails here
```
 
### What did you expect to see?

Some indication of success 
 

### What did you see instead?
```
sudo make install            
[INSTALL] Installing public C headers
[MAKE]    Generating cache.mk
[STRIP]   Stripping libgpr.a
[STRIP]   Stripping libgrpc.a
[STRIP]   Stripping libgrpc_cronet.a
[STRIP]   Stripping libgrpc_unsecure.a
[INSTALL] Installing C pkg-config files
[INSTALL] Installing libgpr.a
[INSTALL] Installing libgrpc.a
[INSTALL] Installing libgrpc_cronet.a
[INSTALL] Installing libgrpc_unsecure.a
[STRIP]   Stripping libgpr.so.4.0.0-dev
[STRIP]   Stripping libgrpc.so.4.0.0-dev
[STRIP]   Stripping libgrpc_cronet.so.4.0.0-dev
[STRIP]   Stripping libgrpc_unsecure.so.4.0.0-dev
[INSTALL] Installing libgpr.so.4.0.0-dev
[INSTALL] Installing libgrpc.so.4.0.0-dev
[INSTALL] Installing libgrpc_cronet.so.4.0.0-dev
[INSTALL] Installing libgrpc_unsecure.so.4.0.0-dev
[INSTALL] Installing public C++ headers
[HOSTLD]  Linking /home/anton/git/grpc/bins/opt/grpc_cpp_plugin
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::get_output_type_name[abi:cxx11]() const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:55: undefined reference to `google::protobuf::MethodDescriptor::output_type() const'
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::get_input_type_name[abi:cxx11]() const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:52: undefined reference to `google::protobuf::MethodDescriptor::input_type() const'
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::output_type_name[abi:cxx11]() const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:48: undefined reference to `google::protobuf::MethodDescriptor::output_type() const'
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::input_type_name[abi:cxx11]() const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:45: undefined reference to `google::protobuf::MethodDescriptor::input_type() const'
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::get_module_and_message_path_output(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:71: undefined reference to `google::protobuf::MethodDescriptor::output_type() const'
/home/anton/git/grpc/objs/opt/src/compiler/cpp_plugin.o: In function `ProtoBufMethod::get_module_and_message_path_input(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) const':
/home/anton/git/grpc/./src/compiler/protobuf_plugin.h:62: undefined reference to `google::protobuf::MethodDescriptor::input_type() const'
collect2: error: ld returned 1 exit status
Makefile:15035: recipe for target '/home/anton/git/grpc/bins/opt/grpc_cpp_plugin' failed
make: *** [/home/anton/git/grpc/bins/opt/grpc_cpp_plugin] Error 1
```

 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).
 
### Anything else we should know about your project / environment?
Yes.
Running `make` also fails!

```
[CXX]     Compiling src/cpp/server/server_posix.cc
[CXX]     Compiling src/cpp/thread_manager/thread_manager.cc
[CXX]     Compiling src/cpp/util/byte_buffer_cc.cc
[CXX]     Compiling src/cpp/util/slice_cc.cc
[CXX]     Compiling src/cpp/util/status.cc
[CXX]     Compiling src/cpp/util/string_ref.cc
[CXX]     Compiling src/cpp/util/time_cc.cc
[HOSTCXX] Compiling src/compiler/cpp_generator.cc
[HOSTCXX] Compiling src/compiler/csharp_generator.cc
[HOSTCXX] Compiling src/compiler/node_generator.cc
[HOSTCXX] Compiling src/compiler/objective_c_generator.cc
[HOSTCXX] Compiling src/compiler/php_generator.cc
In file included from src/compiler/php_generator.cc:23:0:
./src/compiler/php_generator_helpers.h: In function ‘grpc::string grpc_php_generator::GetPHPServiceFilename(const FileDescriptor*, const ServiceDescriptor*, const string&)’:
./src/compiler/php_generator_helpers.h:51:23: error: ‘const class google::protobuf::FileOptions’ has no member named ‘has_php_namespace’; did you mean ‘has_csharp_namespace’?
   if (file->options().has_php_namespace()) {
                       ^~~~~~~~~~~~~~~~~
./src/compiler/php_generator_helpers.h:52:39: error: ‘const class google::protobuf::FileOptions’ has no member named ‘php_namespace’; did you mean ‘csharp_namespace’?
     oss << ReplaceAll(file->options().php_namespace(), "\\", "/");
                                       ^~~~~~~~~~~~~
src/compiler/php_generator.cc: In function ‘grpc::string grpc_php_generator::GenerateFile(const FileDescriptor*, const ServiceDescriptor*, const string&)’:
src/compiler/php_generator.cc:153:25: error: ‘const class google::protobuf::FileOptions’ has no member named ‘has_php_namespace’; did you mean ‘has_csharp_namespace’?
     if (file->options().has_php_namespace()) {
                         ^~~~~~~~~~~~~~~~~
src/compiler/php_generator.cc:154:39: error: ‘const class google::protobuf::FileOptions’ has no member named ‘php_namespace’; did you mean ‘csharp_namespace’?
       php_namespace = file->options().php_namespace();
                                       ^~~~~~~~~~~~~
Makefile:2525: recipe for target '/home/anton/git/grpc/objs/opt/src/compiler/php_generator.o' failed
make: *** [/home/anton/git/grpc/objs/opt/src/compiler/php_generator.o] Error 1
```
Unless I run `make HAS_SYSTEM_PROTOBUF=false`.

But if then run `make HAS_SYSTEM_PROTOBUF=false`, while make succeeds, then `make install` will fail. It all seems rather hopeless.

**Comment by nicolasnoble** [comment]
You can do `make install HAS_SYSTEM_PROTOBUF=false`. This seems like you need the latest version of protobuf installed on your system however.

**Comment by magick93** [comment]
Yes, I have tried this. And yes I do have the latest version installed.

Im running into so many issues I must admit I am getting confused and losing tack.

**Comment by julianoes** [comment]
I have the same issue on Fedora 25:

```
gcc --version
gcc (GCC) 7.1.1 20170622 (Red Hat 7.1.1-3)
```

```
protoc --version
libprotoc 3.2.0
```


```
make -j8                                                                                                                                                                          <<<
[MAKE]    Generating cache.mk
[C]       Compiling src/core/ext/transport/cronet/transport/cronet_transport.c
[HOSTCXX] Compiling src/compiler/php_generator.cc
src/core/ext/transport/cronet/transport/cronet_transport.c: In function ‘create_grpc_frame’:
src/core/ext/transport/cronet/transport/cronet_transport.c:653:10: warning: conversion to ‘uint8_t {aka unsigned char}’ from ‘int’ may alter its value [-Wconversion]
   *p++ = (flags & GRPC_WRITE_INTERNAL_COMPRESS) ? 1 : 0;
          ^
[AR]      Creating /home/julianoes/src/grpc/libs/opt/libgrpc_cronet.a
[LD]      Linking /home/julianoes/src/grpc/libs/opt/libgrpc_cronet.so.4.0.0
In file included from src/compiler/php_generator.cc:23:0:
./src/compiler/php_generator_helpers.h: In function ‘grpc::string grpc_php_generator::GetPHPServiceFilename(const FileDescriptor*, const ServiceDescriptor*, const string&)’:
./src/compiler/php_generator_helpers.h:51:23: error: ‘const class google::protobuf::FileOptions’ has no member named ‘has_php_namespace’; did you mean ‘has_csharp_namespace’?
   if (file->options().has_php_namespace()) {
                       ^~~~~~~~~~~~~~~~~
                       has_csharp_namespace
./src/compiler/php_generator_helpers.h:52:39: error: ‘const class google::protobuf::FileOptions’ has no member named ‘php_namespace’; did you mean ‘csharp_namespace’?
     oss << ReplaceAll(file->options().php_namespace(), "\\", "/");
                                       ^~~~~~~~~~~~~
                                       csharp_namespace
src/compiler/php_generator.cc: In function ‘grpc::string grpc_php_generator::GenerateFile(const FileDescriptor*, const ServiceDescriptor*, const string&)’:
src/compiler/php_generator.cc:153:25: error: ‘const class google::protobuf::FileOptions’ has no member named ‘has_php_namespace’; did you mean ‘has_csharp_namespace’?
     if (file->options().has_php_namespace()) {
                         ^~~~~~~~~~~~~~~~~
                         has_csharp_namespace
src/compiler/php_generator.cc:154:39: error: ‘const class google::protobuf::FileOptions’ has no member named ‘php_namespace’; did you mean ‘csharp_namespace’?
       php_namespace = file->options().php_namespace();
                                       ^~~~~~~~~~~~~
                                       csharp_namespace

```

To get this far, I had to disable `-Werror` because of `-Wconversion` and `-Wimplicit-fallthrough`.

**Comment by julianoes** [comment]
The same happens in Ubuntu 16.04 in docker
```
In file included from src/compiler/php_generator.cc:23:0:
./src/compiler/php_generator_helpers.h: In function 'grpc::string grpc_php_generator::GetPHPServiceFilename(const FileDescriptor*, const ServiceDescriptor*, const string&)':
./src/compiler/php_generator_helpers.h:51:23: error: 'const class google::protobuf::FileOptions' has no member named 'has_php_namespace'
   if (file->options().has_php_namespace()) {
                       ^
./src/compiler/php_generator_helpers.h:52:39: error: 'const class google::protobuf::FileOptions' has no member named 'php_namespace'
     oss << ReplaceAll(file->options().php_namespace(), "\\", "/");
                                       ^
src/compiler/php_generator.cc: In function 'grpc::string grpc_php_generator::GenerateFile(const FileDescriptor*, const ServiceDescriptor*, const string&)':
src/compiler/php_generator.cc:153:25: error: 'const class google::protobuf::FileOptions' has no member named 'has_php_namespace'
     if (file->options().has_php_namespace()) {
                         ^
src/compiler/php_generator.cc:154:39: error: 'const class google::protobuf::FileOptions' has no member named 'php_namespace'
       php_namespace = file->options().php_namespace();
                                       ^
Makefile:2525: recipe for target '/grpc/objs/opt/src/compiler/php_generator.o' failed
make: *** [/grpc/objs/opt/src/compiler/php_generator.o] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by julianoes** [comment]
This seems to be the same error as #12316.

**Comment by julianoes** [comment]
This can be fixed using suggestion from #12316:
```
make HAS_SYSTEM_PROTOBUF=false
```
